### DISCUSSION TITLE
Completely static components

### DISCUSSION THREAD

**Discussion by thmsmlr** [vercel/next.js#14840] [Ideas]
Hey next.js community. I'm a big fan of the project. I've been following and using it for a few years. I've recently started to increasingly develop with next.js 9.3+ due to the more advanced features with API routes and static generation. But i've run into some issues, i'm not sure what to search for on the forums, so excuse me if this has already been discussed.

I have found myself routinely running into issues regarding the client/server side split of isomorphic code. When building a page as far as I can tell there are 2-3 different versions of the page that are generated.

1. Static
2. Server Rendered
3. Client Rendered

And then there are transitions between those pages. For instance, when a static page is generated, it'll first display the HTML that was generated at build time, but then after the javascript loads, it'll rerender the page client side with whatever the tree-shaked, minified version of the code is (not sure on the details, just that the code is observably different).

This has caused some issues for me.

Most recently, i've been working on my personal blog. I decided to SSG the blogs using dynamic routes powered by some pages written in [Notion](https://github.com/splitbee/react-notion/issues/15). This page uses `react-notion` which in turn uses `prism.js` to syntax highlight code blocks.

The issue that I was observing was that when I client side transitioned to a blog post, the code would render properly, and when I server rendered with javascript disabled the page would properly render. However, when I server rendered with javascript enabled, upon the react rehydration and rerender, it would try to reparse the code resulting in a different highlighting.

The ultimate reason for this once I debugged is that when the server rendering was generated at build time, it had access to all the code available in the `node_modules` however, when the client side rehydrated, the `prism.js` library minified out all the non-default language bindings causing the client side code to be unable to parse the correct language. This would result in a quick flicker between highlighted code to a non-highlighted code block. The solution was to simply explicitly load the specific language packages that I need. But nevertheless was an interesting bug that got me thinking.

In my case I was writing a blog post, the post didn't need to rehydrate and render after static page load. There was nothing dynamic on the page. I would've been sufficient to just fully render it statically and remove all the javascript. This would've solved my issue because it doesn't matter how `prism.js` gets packaged, it only matters what happens at build time. However I understand that next.js is suppose to be progressive framework. So it can't determine whether or not it needs to rehydrate. Suppose I had a react hook in the page to update a counter or something. Immediately I would've needed the rehydrate. How would you differentiate between the two usecases automatically?

So this got me thinking. Maybe there is some kind of babel plugin magic or what not that could detect during the build of a static page whether or not there is any case in which the page would need to rerender after the static page was loaded and then optimize to not load any javascript at all.

However, even if this was possible, it'd probably be confusing from a user perspective and error prone. So that got me thinking of whether the page API could be extended to determine whether a specific page should be completely static, or static + rehydration. As the Pythonistas say, explicit is better than implicit. So surely that was a good idea right?

I was sitting on this for a few hours and realized it was kind of stupid, because even if you were to analyze my usecase. The feature would immediately break down the second that I added a fancy JS enabled dropdown to my navigation bar. So basically it'd only be an optimization for the most basic of cases, and that's probably not that useful for the community.

That's when I remember the dynamic components feature of next.js where you can dynamically load a component and mark it as client side only. I figured this is probably the exact API that you would want. I don't want to send all of `prism.js` to my users, I know for a fact that the code blocks in my component tree are never going to be rerender. What if I could mark that component as `static_only` or something like that which `next.js` knew to modify into a `<div dangerouslySetInnerHTML=... />` component or something like that.

Then you wouldn't need to know when to rehydrate or not, you can still be very progressive as an framework without confusing the users, but those expert users who know for certain certain components in the tree will never be rerendered can prevent the javascript for that component from ever being sent to the browser.

If you've gotten to this point, thanks for reading my ramblings. What are your thoughts?

**Comment by leerob** [comment]
> So that got me thinking of whether the page API could be extended to determine whether a specific page should be completely static, or static + rehydration.

This has been discussed and there's an experimental flag for disabling runtime JS!

https://github.com/vercel/next.js/pull/11949
https://github.com/vercel/next.js/issues/5054

**Reply by thmsmlr** [reply]
Ah interesting, are there any discussions about only server rendering parts of the component tree. Or is that just inherently a bad idea? 

I'm thinking for the usecase where you have some expensive JS, or big dependency for something that you know is completely static. You could mark the component as static and it'll only render once at SSR or SSG time, then the client won't get the javascript to rerender, but instead will just receive a component that statically sets the inner HTML?

**Reply by Siggnja** [reply]
@thmsmlr We have been thinking about this as-well. One thing we tried was to flag components as server side only and using a bundler to recognize the components to remove them from the client bundle. We then use [react-lazy-hydration](https://www.npmjs.com/package/react-lazy-hydration) to avoid the components from hydrating on the client.

The only issue we have had with this method was when doing client side routing, the components obviously aren't shipped to the client so these components are not rendered.

**Reply by solarkraft** [reply]
> We then use [react-lazy-hydration](https://www.npmjs.com/package/react-lazy-hydration) to avoid the components from hydrating on the client.

Does this imply the ability to later hydrate explicitly?

**Reply by Siggnja** [reply]
Not really. We usually just use some sort of events to trigger hydration, e.g. mouseover. You could probably expand upon this implementation to get some sort of callback from the hydration component and trigger it when you want. Anyway, this topic can probably be closed when server-side-components will land.
